components/director/internal/authenticator/middleware.go

```diff
@@ -159,11 +159,23 @@ func (a *Authenticator) getKeyFunc() func(token *jwt.Token) (interface{}, error)
 		switch token.Method.Alg() {
 		case jwt.SigningMethodRS256.Name:
 			a.mux.Lock()
-			keys := a.cachedJWKs.Keys
+			keys := a.cachedJWKs
 			a.mux.Unlock()
```

Comment:
Not a new problem, but i think the mutex here should be RWMutext.

Here you should lock for read and unlock read. And in `SynchronizeJWKS` methond you should lock/unlock for writing. This way you can have multiple readers at a time, but when a writer comes they will be blocked and the JWKs will be resynced. 

Currently with this lock readers also wait for each other which is uneccesary.
The time that readers are locked is minimized because a reference is stored here and the sync method creates new keys and not modify them. In general looks like a hacky and error prone way.

I would suggest to add a RWMutex lock the whole `SynchronizeJWKS` method for write and the whole method here for read this way the lock will be more robust if the `SynchronizeJWKS` function is changed to not fully replace the jwks but modify it and also the readers will not wait for each other for no reason.